git/mktag.go

```diff
@@ -0,0 +1,145 @@
+	// "\ntype " had to be char 47, since a sha1 is 40 characters long
+	// when written as a hex string.
+	if strings.Index(strval, "\ntype ") != 47 {
+		return Sha1{}, fmt.Errorf(`error: char47: could not find "\ntype "`)
+	}
+	if len(lines) < 3 {
+		return Sha1{}, fmt.Errorf(`error: char48: could not find next "\n"`)
+	}
+	if !strings.HasPrefix(lines[2], "tag ") {
+		pos := len(lines[0]) + len(lines[1]) + 2
+		return Sha1{}, fmt.Errorf(`error: char%d: no "tag " found`, pos)
+	}
+	typ := strings.TrimPrefix(lines[1], "type ")
+	if len(typ) > len("commit") {
+		// "commit" is the longest object type name
+		return Sha1{}, fmt.Errorf(`error: char53: type too long`)
+	}
+	switch typ {
+	case "commit":
+		// FIXME: This should actually verify the object
+	default:
+		return Sha1{}, fmt.Errorf(`error: char7: could not verify object.`)
+	}
+	tagname := strings.TrimPrefix(lines[2], "tag ")
+	for i, c := range tagname {
+		if unicode.IsControl(c) || c == ' ' {
+			// This error is misleading. It did verify it and confirmed that
+			// it was invalid, but the test suite checks for this exact message.
+			//
+			// pos includes 1 for each newline, and 4 for "tag "
+			pos := len(lines[0]) + 1 + len(lines[1]) + 1 + 4 + i + 1
+			return Sha1{}, fmt.Errorf(`error: char%d: could not verify tag name`, pos)
+		}
+	}
+	if !strings.HasPrefix(lines[3], "tagger ") {
+		pos := len(lines[0]) + len(lines[1]) + len(lines[2]) + 3
+		return Sha1{}, fmt.Errorf(`error: char%d: could not find "tagger "`, pos)
+	}
+	tagger := strings.TrimSpace(strings.TrimPrefix(lines[3], "tagger "))
+	if tagger[0] == '<' {
+		pos := len(lines[0]) + len(lines[1]) + len(lines[2]) + 3 + len("tagger ")
+		return Sha1{}, fmt.Errorf(`error: char%d: missing tagger name`, pos)
+	}
+
+	taggerRe := regexp.MustCompile(`(.*)\<(.*)\>(.*)`)
+
+	taggerPieces := taggerRe.FindStringSubmatch(tagger)
+	// 0 = whole match
+	// 1 = author
+	// 2 = email
+	// 3 = time
+	if len(taggerPieces) != 4 {
+		pos := len(lines[0]) + len(lines[1]) + len(lines[2]) + 3 + len("tagger ")
+		return Sha1{}, fmt.Errorf(`error: char%d: malformed tagger field`, pos)
+	}
+
+	if sp := strings.Index(taggerPieces[2], " "); sp >= 0 {
+		pos := len(lines[0]) + len(lines[1]) + len(lines[2]) + 3 + len("tagger ")
+		return Sha1{}, fmt.Errorf(`error: char%d: malformed tagger field`, pos)
+	}
+
+	// It needs to match (\d)+ (\+|\-)\d+, but if the first part doesn't match we need
+	// to return a different error than the second to pass the git test suite, so we
+	// do this in two parts
+	timestampRe := regexp.MustCompile(`^ (\d+) (\+|\-)(\d+)$`)
+	timepieces := timestampRe.FindStringSubmatch(taggerPieces[3])
+	if len(timepieces) == 0 {
+		// This is all really stupid stupid, but we need to ensure that we report
+		// the same char position and error message as git to get the tests to pass.
+
+		// If the first part isn't an integer, say it's missing.
+		timestampRe2 := regexp.MustCompile(`^ (\d+)`)
```

Comment:
These regexes should probably be pulled out into vars so that if they don't compile you find out right away when dgit starts instead of failing only when this code path is run.

Reply:
Ah, right. Not such a great design for ephemeral command-line tools with many subcommands in them.